### PR TITLE
Make the packet number encryption sampling clearer

### LINKED ISSUE #1387: 5.6 Packet Number Protection is hard to follow
It is hard to understand TLS 5.6 Packet Number Protection.

More specifically, this text block:

```
   In sampling the packet ciphertext, the packet number length is
   assumed to be the smaller of the maximum possible packet number
   encoding (4 octets), or the size of the protected packet minus the
   minimum expansion for the AEAD.  For example, the sampled ciphertext
   for a packet with a short header can be determined by:

   "sample_offset = min(1 + connection_id_length + 4, packet_length -
   aead_expansion) sample =
   packet[sample_offset..sample_offset+sample_length] "
```

It is ambiguous that what packet_length is.  Is it the Packet length field in Long packet header?  But it does not appear in Short header.  Is it the length of (Protected) Payload?  Or the length of QUIC packet, or UDP packet?

The text says "the packet number length is ...", but the example seems to be offset to the byte to sample.

It would be also nice to show the example of Long packet.

### PR DESCRIPTION
**Pull request by martinthomson**
Fixes #1387.

### CODE REVIEW

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -942,16 +942,28 @@ Packet number protection is applied after packet protection is applied (see
...
 
+A packet with a long header is sampled in the same way, noting that multiple
+QUIC packets might be included in the same UDP datagram and that each one is
+handled separately.
+
+~~~
+sample_offset = min(2 + len(destination_connection_id) +
+                      len(source_connection_id) +
+                      len(payload_length) + 4,
+                    packet_length - aead_expansion)
```
Comment (nibanks) [inline draft-ietf-quic-tls.md]: I assume `packet_length` refers to the QUIC packet length and not the UDP payload length, right? If so, how's this supposed to be determined from a long header packet, especially if it's the first of several compound packets?

Comment (martinthomson) [reply draft-ietf-quic-tls.md]: Ahh, I see the problem.  You don't know where the payload length starts.  It seems like we need to make the payload length cover the packet number length.  More changes inbound.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -942,16 +942,28 @@ Packet number protection is applied after packet protection is applied (see
...
                     packet_length - aead_expansion)
 sample = packet[sample_offset..sample_offset+sample_length]
 ~~~
 
+A packet with a long header is sampled in the same way, noting that multiple
+QUIC packets might be included in the same UDP datagram and that each one is
+handled separately.
+
+~~~
+sample_offset = min(2 + len(destination_connection_id) +
```
Comment (kazuho) [inline draft-ietf-quic-tls.md]: Shouldn't `2` be `6`? There are the type octet, version (4 octets), CILs (1 octets).

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -942,16 +942,28 @@ Packet number protection is applied after packet protection is applied (see
 encryption algorithm.
 
 In sampling the packet ciphertext, the packet number length is assumed to be the
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-tls.md]: No, but I see how this text is confusing, I'll rephrase and we can try again.

File: draft-ietf-quic-tls.md [hunk 4]
```diff
@@ -941,17 +941,34 @@ Packet number protection is applied after packet protection is applied (see
 {{aead}}).  The ciphertext of the packet is sampled and used as input to an
 encryption algorithm.
 
-In sampling the packet ciphertext, the packet number length is assumed to be the
-smaller of the maximum possible packet number encoding (4 octets), or the size
-of the protected packet minus the minimum expansion for the AEAD.  For example,
-the sampled ciphertext for a packet with a short header can be determined by:
+In sampling the packet ciphertext, the packet number length is assumed to be
+either 4 octets (its maximum possible encoded length), unless there is
```
Comment (MikeBishop) [inline draft-ietf-quic-tls.md]: Either implies two options. I'd just drop it.
